src/lib.rs

```diff
@@ -1387,23 +1468,30 @@ impl<'a, 'input: 'a> DoubleEndedIterator for Children<'a, 'input> {
 }
 
 /// Iterator over a node and its descendants.
-#[derive(Clone, Debug)]
+#[derive(Clone)]
 pub struct Descendants<'a, 'input> {
     doc: &'a Document<'input>,
-    current: NodeId,
-    until: NodeId,
+    nodes: core::iter::Enumerate<core::slice::Iter<'a, NodeData<'input>>>,
+    from: usize,
 }
 
 impl<'a, 'input> Descendants<'a, 'input> {
     #[inline]
     fn new(start: Node<'a, 'input>) -> Self {
+        let from = start.id.get_usize();
+
+        let until = start
+            .d
+            .next_subtree
+            .map(NodeId::get_usize)
+            .unwrap_or(start.doc.nodes.len());
+
+        let nodes = start.doc.nodes[from..until].iter().enumerate();
```

Comment:
The main problem I had, was that mapping here is not possible with a closure as the type to embed in `Descendants` would become unnameable otherwise.

(As discussed in https://github.com/RazrFalcon/roxmltree/pull/80#discussion_r1028314732, I think these types should eventually go away completely using something like

```rust
pub type Descendants<'a> = impl ExactSizeIterator + DoubleEndedIterator + Clone + fmt::Debug + 'a;

pub fn descendants(&self) -> Descendants {
  ...

  self.doc.nodes[from..until].iter().enumerate().map(...)
}
```

but that will take some time until that feature is stabilized.)